concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs

```diff
@@ -0,0 +1,717 @@
+markBlockDead :: (MonadState (SkovData pv) m) => BlockHash -> m ()
+markBlockDead blockHash = do
+    blockTable . liveMap . at' blockHash .=! Nothing
+    blockTable . deadBlocks %=! insertDeadCache blockHash
+
+-- |Mark a live block as dead. In addition, purge the block state and maintain invariants in the
+-- transaction table by purging all transaction outcomes that refer to this block.
+-- [Note: this does not affect the branches.]
+markLiveBlockDead ::
+    ( MonadState (SkovData pv) m,
+      BlockStateStorage m,
+      GSTypes.BlockState m ~ PBS.HashedPersistentBlockState pv
+    ) =>
+    BlockPointer pv ->
+    m ()
+markLiveBlockDead bp = do
+    let bh = getHash bp
+    markBlockDead bh
+    purgeBlockState $ bpState bp
+    mapM_ (markTransactionDead bh) (blockTransactions bp)
+
+-- |Mark a block as pending in the block table.
+-- (Note, this does not update the pending block table.)
+markPending :: (MonadState (SkovData pv) m) => PendingBlock -> m ()
+markPending pb = blockTable . liveMap . at' (getHash pb) ?=! MemBlockPending pb
+
+-- * Operations on pending blocks
+
+-- $pendingBlocks
+-- Pending blocks are conceptually stored in a min priority search queue,
+-- where multiple blocks may have the same key, which is their parent,
+-- and the priority is the block's round number.
+-- When a block arrives (possibly dead), its pending children are removed
+-- from the queue and handled.  This uses 'takePendingChildren'.
+-- When a block is finalized, all pending blocks with a lower or equal round
+-- number can be handled (they will become dead, since they can no longer
+-- join the tree).  This uses 'takeNextPendingUntil'.
+
+-- |Add a block to the pending block table and queue.
+-- (Note, this does not update the block table.)
+addPendingBlock :: (MonadState s m, HasPendingBlocks s) => PendingBlock -> m ()
+addPendingBlock sb = do
+    pendingBlocksQueue %= MPQ.insert theRound (blockHash, parentHash)
+    pendingBlocksTable . at' parentHash . non [] %= (sb :)
+  where
+    blockHash = getHash sb
+    theRound = blockRound sb
+    parentHash = blockParent sb
+
+-- |Take the set of blocks that are pending a particular parent from the pending block table.
+-- Note: this does not remove them from the pending blocks queue; blocks should be removed from
+-- the queue as the finalized round progresses.
+takePendingChildren :: (MonadState s m, HasPendingBlocks s) => BlockHash -> m [PendingBlock]
+takePendingChildren parent = pendingBlocksTable . at' parent . non [] <<.= []
+
+-- |Return the next block that is pending its parent with round number
+-- less than or equal to the given value, removing it from the pending
+-- table.  Returns 'Nothing' if there is no such pending block.
+takeNextPendingUntil :: (MonadState s m, HasPendingBlocks s) => Round -> m (Maybe PendingBlock)
+takeNextPendingUntil targetRound = takeNextUntil =<< use pendingBlocksQueue
+  where
+    takeNextUntil ppq = case MPQ.minViewWithKey ppq of
```

Comment:
Renamed to pbq. It's the pendingBlocksQueue we're binding into `takeNextUntil` in the line just above.
Unfortunately we cannot name it much better because of how haskell works. 

I don't think a `pendingBlksQueue` would be better either.